storybookjs/storybook
# UI: Add an experimental API for adding sidebar filter functions at runtime

Telescoped off: https://github.com/storybookjs/storybook/pull/23802

## What I did

I introduced a new manager API!
It's marked as experimental for now.

The purpose is to allow users and addon-creators to inject (dynamically at runtime) filter functions that filter the list of stories in the sidebar. I know this has been a feature that has been requested many times for a varied use-cases.

I've implemented the actual filtering in the sidebar, but this code will likely need optimization.
And with this optimization there will be assumptions in when the filters are re-evaluated.

A `FilterFunction` receives an (with status extended) `IndexHash_Item` and returns a boolean.
Storybook will only ask to filter leaf-items (story and docs).
Storybook will filter out empty components/groups/roots for you.

## How to test

This API can be used via from `manager.ts`:
```ts
import { addons } from '@storybook/manager-api';
import startCase from 'lodash/startCase.js';

addons.setConfig({
  sidebar: {
    filters: {
      a: (item) => item.depth === 2,
    },
  },
});
```

The expected result is that the sidebar stories are filtered.

## Checklist



      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers


`["cleanup", "BREAKING CHANGE", "feature request", "bug", "build", "documentation", "maintenance", "dependencies", "other"]`



### 🦋 Canary release



This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_
